concurrency/NonMemberPreferencesMigrationHelper.java

```diff
@@ -0,0 +1,397 @@
+
+    @NonNull private final ObjectMapper mapper;
+    @NonNull private final File file;
+    @NonNull private final TypeReference<ConcurrentHashMap<String, Object>> typeRef;
+    @NonNull private final CountDownLatch loadedFromDisk;
+    @NonNull private ConcurrentHashMap<String, Object> map;
+    @NonNull NonMemberSettingsMigrationMetricsHelper metricsHelper;
+
+    private Disposable commitToDiskDisposable;
+    private String accountIdSettingsSynchronizationDoneFor;
+
+    public NonMemberPreferencesMigrationHelper(@Nullable final Profiler profiler) {
+        mapper = new ObjectMapper();
+        typeRef = new TypeReference<ConcurrentHashMap<String, Object>>() { };
+        file = new File(String.format(ABS_PATH_FORMAT,
+                Environment.getExternalStorageDirectory().getAbsolutePath()));
+        loadedFromDisk = new CountDownLatch(1);
+        map = new ConcurrentHashMap<>();
+        metricsHelper = new NonMemberSettingsMigrationMetricsHelper(profiler);
+
+        loadFromDisk();
+    }
+
+    @VisibleForTesting
+    public NonMemberPreferencesMigrationHelper() {
+        this(null);
+    }
+
+    /**
+     * Retrieve a String value from the persistent preferences.
+     * @param key The name of the preference to retrieve.
+     * @param defaultValue Value to return if this preference does not exist.
+     * @return Returns the preference value if it exists, or defaultValue. Throws
+     * ClassCastException if there is a preference with this name that is not
+     * a String.
+     *
+     * @throws ClassCastException If setting value's type is not String.
+     */
+    @Nullable
+    public String getString(@NonNull final String key, @Nullable final String defaultValue) {
+        awaitLoadFromDisk();
+        return map.containsKey(key) ? (String)map.get(key) : defaultValue;
+    }
+
+    /**
+     * Set a String value in the persistent preferences store.
+     * @param key The name of the preference to modify.
+     * @param value The new value for the preference.
+     */
+    public void putString(@NonNull final String key, @NonNull final String value) {
+        awaitLoadFromDisk();
+        map.put(key, value);
+        commitToDisk();
+    }
+
+    /**
+     * Retrieve a boolean value from the persistent preferences.
+     *
+     * @param key The name of the preference to retrieve.
+     * @param defaultValue Value to return if this preference does not exist.
+     *
+     * @return Returns the preference value if it exists, or defaultValue. Throws
+     * ClassCastException if there is a preference with this name that is not
+     * a boolean.
+     *
+     * @throws ClassCastException If setting value's type is not Boolean.
+     */
+    @Nullable
+    public Boolean getBoolean(@NonNull final String key, @Nullable final Boolean defaultValue) {
+        awaitLoadFromDisk();
+        return map.containsKey(key) ? (Boolean) map.get(key) : defaultValue;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **120**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The remove that can remove the key is at line: **144**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).